src/utils/uintarray-encoding.js

```diff
@@ -4,31 +4,157 @@
 // @flow
 
 /**
- * Space-efficient url component compatible encoding for arrays of 32bit
- * unsigned integers. Smaller numbers take up fewer characters.
+ * Space-efficient url component compatible encodings for arrays of 32bit
+ * unsigned integers.
  */
 
-const encodingChars: string =
+// Depending on who you ask, there are different sets of allowed (non-reserved)
+// characters you can use in a URL component without percent-encoding. The most
+// stringent set is defined by RFC 3986, which only allows 66 characters:
+// 10 digits + 26 * 2 letters + 4 special characters . _ - ~
+// The JavaScript encodeURIComponent function is less strict than RFC 3986 and
+// also allows the following five characters, but we do not use them: ! ' ( ) *
+//
+// This file implements two encodings for arrays of unsigned 32 integers:
+// 'consecutive-12' and 'compact-64'.
+//
+// consecutive-12
+// --------------
+//
+// The consecutive-12 encoding makes use of the following 12 characters: 0123456789-_
+// Numbers are written out normally (as decimal numbers), separated by -, with a special
+// a_b syntax which means "all consecutive numbers from a to b (including b)".
+// This encoding is useful for lists of numbers where it's common to have multiple
+// consecutive numbers in a row, e.g. [1, 2, 3, 4, 5, 9, 10, 11, 12].
+// Examples:
+//  [0] -> "0", [9, 10] -> "9_10", [31, 167, 32, 33, 34] -> "31-167-32_34"
+//
+// compact-64
+// ----------
+//
+// The compact-64 encoding makes use of 64 of the allowed 66 characters:
+// 0123456789abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ._
+// It leaves - and ~ available for other uses.
+// This encoding is useful for lists of numbers in arbitrary orders. Individual numbers
+// take up fewer characters than in the decimal encoding, and there is no need for
+// separator characters.
+// Examples:
+//  [0] -> "0", [9, 10] -> "9a", [31, 167, 32, 33, 34] -> "vB7x0x1x2"
+//
+// Here's how it works: The 64 possible characters give us 6 bits per character (2^6 == 64).
+// We split each 32-bit number into 5-bit segments. Each of these 5-bit segments becomes
+// one character. This leaves 1 bit per character for a "Are there more characters
```

Comment:
I was wondering the difference with base64, but reading the code I see that we encode only the significant bits, and not the leading zeros. Can you please just clarify this here? (`We split each 32-bit number into 5-bit segments` sounds like we always have the same count of segments)
That's what you wrote previously with `Smaller numbers take up fewer characters.`